### PR TITLE
Plumb through unique_ptr use around llvm::Module.

### PR DESCRIPTION
**Pull request by zvookin**
A few other C++11 cleanups as well.

### CODE REVIEW

File: src/CodeGen_LLVM.cpp [hunk 1]
```diff
@@ -31,11 +31,9 @@
 
 namespace Halide {
 
-llvm::Module *codegen_llvm(const Module &module, llvm::LLVMContext &context) {
-    Internal::CodeGen_LLVM *cg = Internal::CodeGen_LLVM::new_for_target(module.target(), context);
-    llvm::Module *out = cg->compile(module);
-    delete cg;
-    return out;
+std::unique_ptr<llvm::Module> codegen_llvm(const Module &module, llvm::LLVMContext &context) {
+    std::unique_ptr<Internal::CodeGen_LLVM> cg(Internal::CodeGen_LLVM::new_for_target(module.target(), context));
```
Comment (dsharletg) [inline src/CodeGen_LLVM.cpp]: Maybe new_for_target should return std::unique_ptr?

File: src/JITModule.cpp [hunk 2]
```diff
@@ -153,26 +150,30 @@ char *start, *end;
 #endif
 
 // Retrieve a function pointer from an llvm module, possibly by compiling it.
-JITModule::Symbol compile_and_get_function(ExecutionEngine *ee, llvm::Module *mod, const string &name, bool optional = false) {
+#if 0
```
Comment (dsharletg) [inline src/JITModule.cpp]: Unconditionally dead code?

File: src/LLVM_Runtime_Linker.cpp [hunk 3]
```diff
@@ -376,18 +372,19 @@ void link_modules(std::vector<llvm::Module *> &modules, Target t) {
...
                                                 llvm::Linker::DestroySource, &err_msg);
-        #endif
+            #endif
 
-        // No matter which version about ran, we passed ownership of
-        // the second module to LLVM. modules[i] is now a dangling
-        // pointer. null it out so we don't accidentally use it.
-        modules[i] = nullptr;
+            // LLVM consumed the module so reset the unique_ptr.
+            modules[i].reset();
```
Comment (abadams) [inline src/LLVM_Runtime_Linker.cpp]: Not at all confident in my understanding of std::move, but if llvm consumed the module, shouldn't this be a no-op? I.e. if something move-constructed from the module the unique ptr should be already reset.

File: src/LLVM_Runtime_Linker.cpp [hunk 4]
```diff
@@ -408,50 +405,45 @@ void link_modules(std::vector<llvm::Module *> &modules, Target t) {
...
-    for (llvm::Module::global_iterator iter = module->global_begin(); iter != module->global_end(); iter++) {
-        if (llvm::GlobalValue *gv = llvm::dyn_cast<llvm::GlobalValue>(iter)) {
-            // No variables are part of the public interface (even the ones labelled halide_)
-            llvm::GlobalValue::LinkageTypes t = gv->getLinkage();
-            if (t == llvm::GlobalValue::WeakAnyLinkage) {
-                gv->setLinkage(llvm::GlobalValue::LinkOnceAnyLinkage);
-            } else if (t == llvm::GlobalValue::WeakODRLinkage) {
-                gv->setLinkage(llvm::GlobalValue::LinkOnceODRLinkage);
-            }
+    //    for (llvm::Module::global_iterator iter = module->global_begin(); iter != module->global_end(); iter++) {
```
Comment (abadams) [inline src/LLVM_Runtime_Linker.cpp]: dead code in comment

File: src/CodeGen_LLVM.cpp (hunk 1 above)
Comment (steven-johnson) [inline src/CodeGen_LLVM.cpp]: +1

File: src/CodeGen_LLVM.cpp [hunk 5]
```diff
@@ -436,9 +432,7 @@ llvm::Module *CodeGen_LLVM::compile(const Module &input) {
     CodeGen_LLVM::optimize_module();
 
     // Disown the module and return it.
-    llvm::Module *m = module;
-    module = NULL;
-    return m;
+    return std::move(module);
```
Comment (steven-johnson) [inline src/CodeGen_LLVM.cpp]: Pretty sure that it's always redundant to use std::move() for a returned value (and in fact, I think Google's build system will warn if you do)

File: src/JITModule.cpp [hunk 6]
```diff
@@ -223,14 +224,14 @@ JITModule::JITModule() {
 JITModule::JITModule(const Module &m, const LoweredFunc &fn,
                      const std::vector<JITModule> &dependencies) {
     jit_module = new JITModuleContents();
-    llvm::Module *llvm_module = compile_module_to_llvm_module(m, jit_module.ptr->context);
+    std::unique_ptr<llvm::Module> llvm_module(compile_module_to_llvm_module(m, jit_module.ptr->context));
```
Comment (steven-johnson) [inline src/JITModule.cpp]: could just use "auto"

File: src/JITModule.cpp [hunk 7]
```diff
@@ -468,8 +472,7 @@ void JITModule::memoization_cache_set_size(int64_t size) const {
 }
 
 bool JITModule::compiled() const {
-    // TODO: Track down all uses and make sure changing this to not include "module != NULL" doesn't break anything.
-  return jit_module.ptr->module != NULL;
+  return jit_module.ptr->execution_engine != NULL;
```
Comment (steven-johnson) [inline src/JITModule.cpp]: drive-by NULL -> nullptr?

File: src/LLVM_Output.cpp [hunk 8]
```diff
@@ -245,40 +241,40 @@ void emit_file(llvm::Module *module, const std::string &filename, llvm::TargetMa
...
 
-    pass_manager.run(*module);
+    pass_manager.run(module);
 
     delete target_machine;
 #endif
 }
 
-llvm::Module *compile_module_to_llvm_module(const Module &module, llvm::LLVMContext &context) {
+std::unique_ptr<llvm::Module>compile_module_to_llvm_module(const Module &module, llvm::LLVMContext &context) {
```
Comment (steven-johnson) [inline src/LLVM_Output.cpp]: need space after >

File: src/LLVM_Runtime_Linker.cpp [hunk 9]
```diff
@@ -376,18 +372,19 @@ void link_modules(std::vector<llvm::Module *> &modules, Target t) {
...
+                                                std::move(modules[i]));
         #else
-        bool failed = llvm::Linker::LinkModules(modules[0], modules[i],
+            #if LLVM_VERSION >= 36
+            bool failed = llvm::Linker::LinkModules(modules[0], modules[i]);
+            #else
+            bool failed = llvm::Linker::LinkModules(modules[0], modules[i],
                                                 llvm::Linker::DestroySource, &err_msg);
-        #endif
+            #endif
```
Comment (steven-johnson) [inline src/LLVM_Runtime_Linker.cpp]: indentation

File: src/LLVM_Runtime_Linker.cpp (hunk 3 above)
Comment (steven-johnson) [inline src/LLVM_Runtime_Linker.cpp]: a unique_ptr after std::move is "undefined but valid" -- basically the only things you can (safely) do with one is to assign a new value (or destroy it). resetting here ensures it's in a well-defined state and protects us from nasal demons.

File: src/LLVM_Runtime_Linker.cpp [hunk 10]
```diff
@@ -830,13 +822,13 @@ llvm::Module *get_initial_module_for_ptx_device(Target target, llvm::LLVMContext
 
 
 
-    return modules[0];
+    return std::move(modules[0]);
```
Comment (steven-johnson) [inline src/LLVM_Runtime_Linker.cpp]: should be unnecessary

File: src/LLVM_Runtime_Linker.cpp [hunk 11]
```diff
@@ -848,7 +840,7 @@ llvm::Module *get_initial_module_for_renderscript_device(Target target, llvm::LL
     m->setDataLayout(&dl);
     #endif
 
-    return m;
+    return std::move(m);
```
Comment (steven-johnson) [inline src/LLVM_Runtime_Linker.cpp]: should be unnecessary

File: src/LLVM_Runtime_Linker.cpp (hunk 3 above)
Comment (zvookin) [inline src/LLVM_Runtime_Linker.cpp]: Actually the code here is reseting it because prior to the latest LLVM, the LinkModules call took llvm::Module *, not std_unique_ptr<llvm::Module>. Of course there were some missing gets up above this as well. The joys of ifdefy code.

File: src/CodeGen_LLVM.cpp (hunk 5 above)
Comment (zvookin) [inline src/CodeGen_LLVM.cpp]: This is returning a field of the class and the "try treating it as an r-value first" thing doesn't apply to members of a class. I'll see if I can remove some of the others. (I put them in as I was getting compile errors here. I just went and read the spec. I know more than I did before but I do not feel enlightened.) Another approach would be to not make the module a member of this class, but I'm not sure that is right either.

File: src/LLVM_Runtime_Linker.cpp (hunk 3 above)
Comment (abadams) [inline src/LLVM_Runtime_Linker.cpp]: maybe release() is the right thing for the old llvms rather than get() - they still consume the module, so I don't think it's safe for us to destroy.

File: src/JITModule.cpp (hunk 6 above)
Comment (zvookin) [inline src/JITModule.cpp]: I'd rather wait to start pushing more use of unique_ptr through Halide. I think it is a good idea, but I also think it could happen in a fair number of places and would rather undertake that as a specific task.

File: src/JITModule.cpp (hunk 7 above)
Comment (zvookin) [inline src/JITModule.cpp]: Was there some reason nullptr couldn't happen yet?

Comment (abadams) [inline src/JITModule.cpp]: Not that I can recall. It seems to be used in some places without causing build failures.

File: src/LLVM_Runtime_Linker.cpp (hunk 10 above)
Comment (zvookin) [inline src/LLVM_Runtime_Linker.cpp]: Seems to be required as it isn't returning a simple temporary.

File: src/JITModule.cpp (hunk 7 above)
Comment (zvookin) [inline src/JITModule.cpp]: Opened an issue to take care of it globally: https://github.com/halide/Halide/issues/996 .

File: src/CodeGen_LLVM.cpp (hunk 1 above)
Comment (zvookin) [inline src/CodeGen_LLVM.cpp]: I'd rather wait to start pushing more use of unique_ptr through Halide. I think it is a good idea, but I also think it could happen in a fair number of places and would rather undertake that as a specific task.
